### PR TITLE
implement tf.idf similarity 

### PR DESCRIPTION
**Pull request by blanche**
to find similar job description from test data in the corpus

i.e. it can tell you that document X is 82% similar to document Y which will be used as weights for calculating the salary

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -72,3 +77,10 @@
     # get top k terms with the highest idf score (this is atm not very useful
     # but it will be used for the tf.idf similarity stuff
     top_keywords = get_top_idf_features(train_description_feature, 5)
+
+    # calculate similarity between train an testset job descriptions
```
Comment (utkuozbulak) [inline src/main.py:81]: I would refrain from using test set because we probably wont get the results for tests, still no response from Adzuna

Edit: But I get the motivation behind, we will use x portion of train etc etc. I'm looking at the stuff.

Comment (blanche) [reply src/main.py:81]: good point, lets in include that in #25 ?

File: src/main.py [hunk 2]
```diff
@@ -72,3 +77,10 @@
...
+    # this is of high order complexity - test it on a subset of the data
+    corpus_list = pandas_vector_to_list(train_description_feature)
+    queries_list = pandas_vector_to_list(test_description_feature)
+    similarity = get_tfidf_similarity(corpus_list, queries_list)
+    relevant_documents = extract_relevant_documents(similarity)
```
Comment (utkuozbulak) [inline src/main.py:86]: For people who don't want to bother with running the code: 
similarity is a (size(train)  size(test)) matrix that has the correlation atings for record to each record.
relevant_documents is the filtered version that is a dictionary, holds any relationship that is more significant than 0.8
Key: 2 value(10, 0.9888 ), meaning second and tenth record's description has huge correlation

Which brings me to the question: @blanche How do you intent to feed this to a machine learning algorithm ? Relevant documents hold the correlation for many to many records and can't be fed as a single feature(column), the simplest way of feeding it is, you have to separate them as:

Correlation_relation_record_to_2 where record is each record(row) and 2 is compared record, but if we use this method we will have 280k features which will blow up the feature set. ( This is identical to adding similarity matrix directly to set )

I get that some descriptions look like other and it is cool that we find it out but I would prefer if you come up with a solution before merging this.
_____________
Some examples from results:
6-26 : Senior Process engineer - junior process engineer
8-109: Chef De Partie needed - Chef De Partie needed, this is literally the same job from same recruiter for same position in same city
20-56: Title Software Engineer - Software Engineer

If the descriptions are almost the same that means the job titles and job category is (almost) the same ( which is). Adding yet another feature that says " hey look these records are similar to each other " will probably do unnecessary emphasis on job titles more than it should.
_____________

TLDR: Nice work, but me wants to see the result in a way we can feed in to ML.

Comment (utkuozbulak) [reply src/main.py:86]: OR ! We can use this as a part of EDA instead of prediction, thats also a way to go.

Comment (blanche) [reply src/main.py:86]: i dont fully know the best way to in cooperate it to be honest, but one simple way is to use it would be as weights in whatever we are going to use.

or as a standalone model: take the average of all salaries proportional to their similarity

Comment (arahayrabedian) [reply src/main.py:86]: would something like clustering work here? where certain similar titles all belong to one 'cluster'?

File: src/main.py (hunk 1 above)
Comment (utkuozbulak) [reply src/main.py:81]: Yarr might do, will look up on it

File: src/main.py (hunk 2 above)
Comment (blanche) [reply src/main.py:86]: yes, this method is regularly used for clustering!
just feed the similarity matrix from (85) in your favourite clustering algorithm
edit: we will want to feed the document vectors, but never mind..

although though clustering you loose information, and i dont quite understand how it would help us here (other than speedup maybe)

Comment (utkuozbulak) [reply src/main.py:86]: @arahayrabedian Might do but not with correlation coefficiency, one thing I can imagine we can do is cluster most common used words for different jobs. I don't know what we will learn from that tbh. 

Imagine for software engineer it is : [software - python - java - full stack - front end - back end - c - rust... ]

For a chef it is : [ cooking - cruise - dessert - meal ... ]

Ok we grouped them, chef words are different than software engineer, what is the insight ?

Comment (arahayrabedian) [reply src/main.py:86]: you have a class to feed in to a learning machine... i.e: it's probably a chef, so it's like other chef jobs --> (hopefully) there should be some sort of correlation between them we can exploit

Comment (blanche) [reply src/main.py:86]: @utku yea i agree
as mentioned, clustering is not what we want here i think.. i.e. we don't just want to throw all software engineers in one pot and as soon as there are in london and are software engineers they get X salary.

using the similarity thing here, you want account for fine nuances and differentiate on a very low level e.g. between a javascript and a php web developer

Comment (blanche) [reply src/main.py:86]: > so it's like other chef jobs --> (hopefully)

hm, that's the point - we can do better than that

Comment (utkuozbulak) [reply src/main.py:86]: Can we have a fave to face discussion on this ? I can't see a feasible way of feeding this to a model and I don't fully understand your suggestions.

**Comment by blanche** [comment]
are there any open issues with this?

**Comment by blanche** [comment]
*bump*

File: src/job_description_feature_extraction.py [hunk 3]
```diff
@@ -1,9 +1,11 @@
 import string
+
+import RAKE
+import numpy as np
```
Comment (arahayrabedian) [inline src/job_description_feature_extraction.py:4]: depending on how strict you wanna be on PEP8 - you may need to reorder your imports.

File: src/job_description_feature_extraction.py [hunk 4]
```diff
@@ -56,6 +60,58 @@ def get_feature_word_count(column_to_count):
...
+    # # use this to overwrite the incoming corpus/queries
+    # corpus = ['bob is in the house', 'susi goes to school', 'et tu brutu']
+    # queries = ['where is the school', 'bob is in the house']
+
+    tfidf_vectorizer = TfidfVectorizer()
+    tfidf_vectorizer.fit(corpus, queries)
+    corpus_vector = tfidf_vectorizer.transform(corpus)
+    queries_vector = tfidf_vectorizer.transform(queries)
+
+    similarity = cosine_similarity(queries_vector, corpus_vector)
```
Comment (arahayrabedian) [inline src/job_description_feature_extraction.py:83]: bruh, this takes like 8 years to run, what's the plan?

File: src/job_description_feature_extraction.py [hunk 5]
```diff
@@ -56,6 +60,58 @@ def get_feature_word_count(column_to_count):
...
+
+    :param similarity: similarity matrix from get_tfidf_similarity()
+    :param sim_threshold: cutoff thershold for relevant document (0.8) by default
+    :return: dict of (document_index, similarity) for every query from similarity matrix
+    """
+    relevant_documents = {}
+    for query_index, similarity_list in enumerate(similarity):
+        for document_index, similarity_score in enumerate(similarity_list):
+            if similarity_score > sim_threshold:
+                relevant_documents[query_index] = (document_index, similarity_score)
```
Comment (arahayrabedian) [inline src/job_description_feature_extraction.py]: this only keeps the 'last seen match' for any query index, should do something like append or pick the max.

```
ipdb> similarity
array([[ 0.51639778,  0.28867513,  0.        ],
       [ 1.        ,  0.        ,  0.        ]])
ipdb> extract_relevant_documents(similarity, sim_threshold=0.5)
{0: (0, 0.51639777949432231), 1: (0, 1.0000000000000002)}
ipdb> extract_relevant_documents(similarity, sim_threshold=0.2)
{0: (1, 0.28867513459481287), 1: (0, 1.0000000000000002)}
```

i.e: we may not get the best result.

**Review by arahayrabedian — CHANGES_REQUESTED** [review]
i'm gonna ignore the discussions we've already had here - just fix up the 'only last seen match' issue and you should be golden.

Dunno about runtimes though, they seem terrifying but that's no fault of yours. (or is it?!)

File: src/job_description_feature_extraction.py (hunk 4 above)
Comment (blanche) [reply src/job_description_feature_extraction.py:83]: 1) leave it running over night, its not our code, can't optimize it really
2) i have a KNN algorithm in the pipeline which should be faster

File: src/job_description_feature_extraction.py (hunk 5 above)
Comment (blanche) [reply src/job_description_feature_extraction.py]: thats a bug, i fixed it in an upcomming PR, downmerge?

Comment (arahayrabedian) [reply src/job_description_feature_extraction.py]: your call, i don't mind, as long as it gets fixed before usage

**Review by arahayrabedian — APPROVED** [review]

**Comment by arahayrabedian** [comment]
golden
